### ISSUE TITLE
Proposal: Reducing the code base by using packages

### ISSUE THREAD

**Issue by machour** [mpdf/mpdf#552]
I've been wondering about ways of reducing the `mpdf/mpdf` code base in order to make it easier to maintain / document.

One way is to extract components with self-contained logic into their own composer packages, or drop them for existing packages.

As an example, `src/Barcode` and `src/QrCode` look generic enough to be extracted to: `mpdf/barcode` & `mpdf/qrcode`.

Those libraries could be used by other projects without requiring `mpdf/mpdf` and could get more ❤️  from the open source community (See #475 for example)

We could also decide to rely on third party librairies, like for example [tecnickcom/tc-lib-barcode](https://github.com/tecnickcom/tc-lib-barcode) which was extracted from TCPDF.

Thoughts?

**Comment by finwe** [comment]
Yes! Also, fonts supporting non-standard dialects could be moved to a package as they use considerable disc space amount.

**Comment by marclaporte** [comment]
For Tiki Wiki CMS Groupware, to be able to fetch packages like mPDF, we have built a PHP web front-end to Composer: https://doc.tiki.org/Packages

And a way to build the packages to make it easy for users to download and install manually via FTP:
https://gitlab.com/tikiwiki/tiki-packages-build

The mPDF project is most welcome to all the ideas and code :-)

**Comment by machour** [comment]
Just saw the two new repositories created. Is this finally happening ? :)
Let me know if I can be of any help.

**Comment by fulldecent** [comment]
QRcode is removed now. Announced in https://github.com/mpdf/mpdf/releases/tag/v8.0.0

**Comment by crazydef** [comment]
I'd rather you went the other way and removed the need to use composer (which I can't use because I'm on a shared server) and simplified the install process to a matter of just uploading all files to a single directory on the web server.  Maybe by packaging features in sub directories that are enabled or disabled simply by their presence or absence.

I tried updating to the latest version earlier and found it pretty much impossible with the server restrictions of my hosting company, so I'm stuck with v6.0 until I can find a suitable alternative.

**Comment by finwe** [comment]
@crazydef No. Composer is de-facto standard now. 

When on a shared server, you can always run composer locally and upload the result yourself.

Maybe better create a separate issue to discuss deployment alternatives/proposals.

**Comment by marclaporte** [comment]
@crazydef 

mPDF is integrated in Tiki Wiki CMS Groupware, which is often installed on shared hosting. So we developed an FTP-friendly solution. It's all Open Source code. I hope it can help others.
https://packages.tiki.org/
https://gitlab.com/tikiwiki/tiki-packages-build
https://doc.tiki.org/Composer
https://doc.tiki.org/mPDF

The other use case are behind-the-firewall installs. Sometimes, it's just easier to work this way, vs dealing with proxies...

**Comment by bby-bishopclark** [comment]
> So we developed an FTP-friendly solution.

It's a great idea, but it does kick the compatibility issue down the field instead of addressing it; and in terms of safety and security in a post-nodeJS-apocalypse more servers are essentially in airgap-but-proxied mode and can only receive their software in common OS package format (deb, rpm, pkg).

Composer may not be compliant, and it's amazing how the big security budget are spent blocking instead of facilitating.  Consider https://git.pld-linux.org/?p=packages/php-mpdf.git
